plugins/aggregate/aggregate.cpp

```diff
@@ -0,0 +1,469 @@
+      if (current_bucket == last_bucket)
+        continue;
+      // Append to chunked array per column in the batch.
+      const auto next_last_row = row + 1;
+      for (int column = 0, builder_column = 0; column < batch->num_columns();
+           ++column) {
+        if (resolved_columns.group_by_actions[column] == action::drop)
+          continue;
+        auto chunks = current_bucket->second[builder_column]->chunks();
+        chunks.emplace_back(
+          batch->column(column)->Slice(last_row, next_last_row - last_row));
+        current_bucket->second[builder_column]
+          = arrow::ChunkedArray::Make(
+              std::move(chunks), current_bucket->second[builder_column]->type())
+              .ValueOr(nullptr);
+        ++builder_column;
+      }
+      last_row = next_last_row;
+    }
+    return caf::none;
+  }
+
+  /// Retrieves the result of the transformation.
+  [[nodiscard]] caf::expected<std::vector<transform_batch>> finish() override {
+    auto result = std::vector<transform_batch>{};
+    result.reserve(resolved_columns_cache_.size());
+    for (const auto& [layout, resolved_columns] : resolved_columns_cache_) {
+      // Modify the layout to drop the columns we don't care about.
+      auto transformations = std::vector<record_type::transformation>{};
+      auto rt = caf::get<record_type>(layout);
+      for (size_t index = 0; const auto& leaf : rt.leaves()) {
+        if (resolved_columns.group_by_actions[index] == action::drop)
+          transformations.push_back({leaf.index, record_type::drop()});
+        ++index;
+      }
+      auto adjusted_rt = rt.transform(std::move(transformations));
+      // TODO: can we handle the error safely in some way?
+      VAST_ASSERT(adjusted_rt);
+      VAST_ASSERT(!layout.has_attributes());
+      auto adjusted_layout = type{layout.name(), *adjusted_rt};
+      auto adjusted_schema = make_arrow_schema(adjusted_layout);
+      // Set up the record batch builder for our aggregated slice.
+      auto batch_builder = std::unique_ptr<arrow::RecordBatchBuilder>{};
+      auto make_batch_builder_result = arrow::RecordBatchBuilder::Make(
+        adjusted_schema, arrow::default_memory_pool(),
+        detail::narrow_cast<int>(resolved_columns.buckets.size()),
+        &batch_builder);
+      VAST_ASSERT(batch_builder->num_fields()
+                  == detail::narrow_cast<int>(adjusted_rt->num_leaves()));
+      VAST_ASSERT(batch_builder->num_fields()
+                  == detail::narrow_cast<int>(
+                    resolved_columns.buckets.begin()->second.size()));
+      // TODO: can we handle the error safely in some way?
+      VAST_ASSERT(make_batch_builder_result.ok());
+      // Iterate over all the buckets to perform the aggregations.
+      for (const auto& bucket : resolved_columns.buckets) {
+        for (int builder_column = 0;
+             auto action : resolved_columns.group_by_actions) {
+          switch (action) {
```

Comment:
I'd like to experiment with additionally doing the aggregation eagerly inside `add(...)` already instead of doing it all at once here in `finish()`. The current way has quite a bit of overhead when applied to a whole partition at once.